environment/core/util/distributions.hpp

```diff
@@ -0,0 +1,474 @@
+
+            friend bool operator==(const param_type& __x, const param_type& __y)
+            {return __x.__a_ == __y.__a_ && __x.__b_ == __y.__b_;}
+            friend bool operator!=(const param_type& __x, const param_type& __y)
+            {return !(__x == __y);}
+        };
+
+    private:
+        param_type __p_;
+
+    public:
+        // constructors and reset functions
+        explicit uniform_int_distribution(result_type __a = 0,
+                                          result_type __b = std::numeric_limits<result_type>::max())
+            : __p_(param_type(__a, __b)) {}
+        explicit uniform_int_distribution(const param_type& __p) : __p_(__p) {}
+        void reset() {}
+
+        // generating functions
+        template<class _URNG> result_type operator()(_URNG& __g)
+        {return (*this)(__g, __p_);}
+        template<class _URNG> result_type operator()(_URNG& __g, const param_type& __p);
+
+        // property functions
+        result_type a() const {return __p_.a();}
+        result_type b() const {return __p_.b();}
+
+        param_type param() const {return __p_;}
+        void param(const param_type& __p) {__p_ = __p;}
+
+        result_type min() const {return a();}
+        result_type max() const {return b();}
+
+        friend bool operator==(const uniform_int_distribution& __x,
+                               const uniform_int_distribution& __y)
+        {return __x.__p_ == __y.__p_;}
+        friend bool operator!=(const uniform_int_distribution& __x,
+                               const uniform_int_distribution& __y)
+        {return !(__x == __y);}
+    };
+
+    template<class _IntType>
+    template<class _URNG>
+    typename uniform_int_distribution<_IntType>::result_type
+    uniform_int_distribution<_IntType>::operator()(_URNG& __g, const param_type& __p)
+    {
+        typedef typename std::conditional<sizeof(result_type) <= sizeof(std::uint32_t),
+            uint32_t, uint64_t>::type _UIntType;
+        const _UIntType _Rp = __p.b() - __p.a() + _UIntType(1);
+        if (_Rp == 1)
+            return __p.a();
+        const std::size_t _Dt = std::numeric_limits<_UIntType>::digits;
+        typedef independent_bits_engine<_URNG, _UIntType> _Eng;
+        if (_Rp == 0)
+            return static_cast<result_type>(_Eng(__g, _Dt)());
+        std::size_t __w = _Dt - __clz(_Rp) - 1;
+        if ((_Rp & (std::numeric_limits<_UIntType>::max() >> (_Dt - __w))) != 0)
+            ++__w;
+        _Eng __e(__g, __w);
+        _UIntType __u;
+        do
+            {
+                __u = __e();
+            } while (__u >= _Rp);
+        return static_cast<result_type>(__u + __p.a());
+    }
+
+    template<class _RealType, std::size_t __bits, class _URNG>
+    _RealType
+    generate_canonical(_URNG& __g)
+    {
+        const std::size_t _Dt = std::numeric_limits<_RealType>::digits;
+        const std::size_t __b = _Dt < __bits ? _Dt : __bits;
+#ifdef _LIBCPP_CXX03_LANG
```

Comment:
This looks safe for me to leave as is undefined anywhere. Just fyi for completeness.

Reply:
I went ahead and took out that branch, since we don't need it anyways.